src/obfuscation/hidden.c

```diff
@@ -0,0 +1,204 @@
+  if (result != 0 || left_length == right_length) {
+    return result;
+  } else {
+    return left_length < right_length ? -1 : 1;
+  }
+}
+
+int HiddenString_Compare(HiddenString *left, HiddenString *right) {
+  UserAndObfuscatedString* l = (UserAndObfuscatedString*)left;
+  UserAndObfuscatedString* r = (UserAndObfuscatedString*)right;
+  return Compare(l->user, l->length, r->user, r->length);
+}
+
+int HiddenString_CompareC(HiddenString *left, const char *right, size_t right_length) {
+  UserAndObfuscatedString* l = (UserAndObfuscatedString*)left;
+  return Compare(l->user, l->length, right, right_length);
+}
+
+int HiddenString_CaseInsensitiveCompareC(HiddenString *left, const char *right, size_t right_length) {
+  UserAndObfuscatedString* l = (UserAndObfuscatedString*)left;
+  return CaseSensitiveCompare(l->user, l->length, right, right_length);
+}
+
+int HiddenString_CaseInsensitiveCompare(HiddenString *left, HiddenString *right) {
+  UserAndObfuscatedString* r = (UserAndObfuscatedString*)right;
+  return HiddenString_CaseInsensitiveCompareC(left, r->user, r->length);
+}
+
+void HiddenString_SaveToRdb(HiddenName* value, RedisModuleIO* rdb) {
+  UserString* text = (UserString*)value;
+  RedisModule_SaveStringBuffer(rdb, text->user, text->length + 1);
+}
+
+int HiddenName_CompareC(const HiddenName *left, const char *right, size_t right_length) {
+  const UserString* l = (const UserString*)left;
+  return Compare(l->user, l->length, right, right_length);
+}
+
+int HiddenName_Compare(const HiddenName* left, const HiddenName* right) {
+  UserString* r = (UserString*)right;
+  return HiddenName_CompareC(left, r->user, r->length);
+}
+
+int HiddenName_CaseInsensitiveCompareC(HiddenName *left, const char *right, size_t right_length) {
+  UserString* l = (UserString*)left;
+  return CaseSensitiveCompare(l->user, l->length, right, right_length);
+}
+
+int HiddenName_CaseInsensitiveCompare(HiddenName *left, HiddenName *right) {
+  UserString* r = (UserString*)right;
+  return HiddenName_CaseInsensitiveCompareC(left, r->user, r->length);
+}
+
+HiddenName *HiddenName_Duplicate(const HiddenName *value) {
+  const UserString* text = (const UserString*)value;
+  return NewHiddenName(text->user, text->length, true);
+}
+
+void HiddenName_TakeOwnership(HiddenName *hidden) {
+  UserString* userString = (UserString*)hidden;
+  userString->user = rm_strndup(userString->user, userString->length);
+}
+
+void HiddenName_Clone(HiddenName* src, HiddenName** dst) {
+  UserString* s = (UserString*)src;
+  if (*dst == NULL) {
+    *dst = NewHiddenName(s->user, s->length, true);
+  } else {
+    UserString* d = (UserString*)*dst;
+    if (s->length > d->length) {
+      d->user = rm_realloc((void*)d->user, s->length);
+      d->length = s->length;
+    }
+    strncpy((void*)d->user, s->user, s->length);
+    if (d->length > s->length) {
+      memset((void*)d->user + s->length, 0, d->length - s->length);
+    }
```

Comment:
Shouldn't we to reallocate `d->user` and update `d->length` also in this case?

Reply:
In the end did add an update for the length.